ggsql announcement (#170)
posit-dev/open-source-website: content/blog/ggsql/alpha_release/index.qmd

---
title: "ggsql: A grammar of graphics for SQL"
date: 2026-04-09
people:
  - Thomas Lin Pedersen
  - Teun Van den Brand
  - George Stagg
  - Hadley Wickham
description: >
  Introducing ggsql, a grammar of graphics for SQL that lets you describe
  visualizations directly inside SQL queries.
image: "ggsql.png"
image-alt: "ggsql logo. A pencil with the body consisting of the standard database silo icon."
categories:
  - Visualization
software:
  - ggsql
languages:
  - SQL
nohero: false
hidesubscription: false
---

Today, we are super excited to formally announce the alpha-release of [ggsql](https://ggsql.org). As the name suggests, ggsql is an implementation of the grammar of graphics based on the SQL syntax, bringing rich, structured visualization support to SQL for the first time.

> In this post we will go over some of the motivations that lead to us developing this tool, as well as give you ample of examples of its use so you can hopefully get as excited about it as we are (if you can't wait to get excited, [jump straight to the examples](#meet-ggsql)).

Reviewer: ```suggestion
In this post we will go over some of the motivations that lead us to develop this tool, as well as give you ample examples of its use; so you can hopefully get as excited about it as we are (if you can't wait to get excited, [jump straight to the examples](#meet-ggsql)).
```

## Why ggsql?

> You may wonder why we have decided to spend time building a tool for SQL rather than pour our time and attention into our R and Python packages. Some of the reasons are:

Reviewer: Here I'd avoid framing it as 'we could have worked on R/python, but instead we didn't'

* We want to engage with and help data analysts and data scientists that predominantly work in SQL
* We want to create an extremely powerful, code-based, visualization tool that doesn't require a runtime
* We have learned so much after 18 years of [ggplot2](https://ggplot2.tidyverse.org/) development and maintenance that it feels right to apply it to a blank canvas
* SQL and the grammar of graphics just fits together extremely well

Let's discuss each in turn.

### Hello, SQL user!

![](greet.png)

> While first R and then Python captured all the attention of the data science revolution, SQL chucked along as the reliant and powerful workhorse beneath it all. We recognize that there are large portions of the people that work with data that does so only or predominantly in SQL. The choice they have for visualizing their data are often suboptimal in our view:

Reviewer: ```suggestion
While first R and then Python captured all the attention of the data science revolution, SQL chucked along as the reliant and powerful workhorse beneath it all. We recognize that there are large portions of people that work with data that do so only or predominantly in SQL. The choice they have for visualizing their data are often suboptimal in our view:
```

* Export the data and use R or Python which may not be within their comfort zone
* Use a GUI-based BI tool with poor support for reproducibility
* Rely on one of the few tools that exist for creating visualizations directly within the query that we feel are not powerful or ergonomic enough

Our goal when designing ggsql was that the syntax should immediately make sense to SQL users, tapping into their expectation of composable, declarative clauses. 

Apart from offering a better way to visualize their data, ggsql is also a way to invite SQL users into our rich ecosystem of code based report generation and sharing build on top of [Quarto](https://quarto.org/).

### No runtime, no problem

> Why does it matter that [ggplot2](https://ggplot2.tidyverse.org/) and [plotnine](https://plotnine.org) requires R and Python installed respectively? Well, even if the situation has improved, R and Python are a much more complicated tech stack than a single small executable. If you spend most of your working day in one of those environment this is not an issue, but if you want to collaborate on a report with someone who doesn't it may well be. 

Reviewer 2: ```suggestion
Why does it matter that [ggplot2](https://ggplot2.tidyverse.org/) and [plotnine](https://plotnine.org) requires R and Python installed respectively? Well, even if the situation has improved, R and Python are a much more complicated tech stack than a single small executable. If you spend most of your working day in one of those environments this is not an issue, but if you want to collaborate on a report with someone who doesn't it may well be. 
```

---

> but if you want to collaborate on a report with someone who doesn't it may well be. 

This part reads a little strange to me, but I don't have an immediate replacement. Perhaps we could expand it a little to describe an example of why it may not be, e.g. a SQL user being overwhelmed with the requirement learn how to drive an entirely new R/Python environment.

More generally, I'm not super convinced by this argument that ggsql is a less complicated tech stack. The counter argument is that either way you need R or Python or the `ggsql` binary/kernel installed to render things. I think of it more as reducing a psychological barrier to visualisation.

The tech-stack story might become more relevant once we have basic interactivity. Then we're comparing to something like Shiny, where you really do need a runtime server-side just to _view_ a dashboard. In the future we might have `ggsql` run in a mode that renders interactive plots to static HTML/JS/JSON that can be deployed to a service without the need for a R/Python runtime. This would be much simpler to deploy than Shiny Server or Connect.

> There is another side to it as well. ggsql code cannot be malicious, in the way Python or R code (or code in any other programming language) can. There is no read-and-upload-all-environment-variables clause, nor an import-dubious-package query in ggsql — it can visualize and that's it. This makes the tool perfect for workflows where you are not the creator of the visualization code (👋 hello, AI agent).

Reviewer: Is saying 'ggsql code cannot be malicious' the same as saying 'people cannot do malicious things with ggsql'?

Reviewer 2: No. Even if we lock things down users can generate large complex queries to DDOS the SQL server.

Reviewer: In that case I think we should clarify and not advertise it as not able to be malicious.

Reviewer 2: I tend to agree, at least in the short term since we don't have strong guardrails yet.

I think the framing should be "better the devil you know". That is, the attack surface of SQL is indeed much smaller than arbitrary R/Python/bash emitted by an LLM, and SQL users can spot dangerous or malicious queries quicker and easier than malicious code (which may be obfuscated).

We, that is ggsql, strives to emit SQL queries that are safe (sidenote: we should have `ggsql audit` or `--dry-run` that emits the queries we plan to send!) and the SQL user can see and audit everything before the `VISUALISE` clause.

Reviewer 2: Database admins also have a lot of control over what a specific database user can or cannot do. I think it is fair to say it is much easier to lockdown/sandbox a database user to some specific read-only tables than it would be to lock down an arbitrary R/Python code running on some Linux box somewhere. Though I know with containers/VMs it's possible.

> ![](malicious.png)

Reviewer 2: We might want to rethink this image. I like the humour but duckdb has things not so far from this.

<img width="407" height="146" alt="Image" src="https://github.com/user-attachments/assets/1d14f832-5fb6-4cc0-8faf-8827c8195fdb" />

Maybe something like a "Spot the issue" image where we show malicious "code" in Python vs SQL. The Python code looks reasonable (e.g. just a typo in a package name?) with blank output. While it's clear that the SQL code is malicious and is blocked.

Or maybe just show the final example in your image, with the error changed to "Access denied: are you crazy?", emphasising that the malicious code was _blocked_ by the database user access controls.

> You may think we have had to swallow some painful bullets by moving away from an interpreted language, but it has also given us a lot. Most importantly, the rigid structure means that we can execute the whole data pipeline as a single SQL query per layer on the backend. This means that if you want to create a bar plot of 10 billion transactions you only ever fetch the count values for each bar from your data warehouse, not the 10 billion rows of data. The same is true for more complicated layer types such as boxplots and density plots.

Reviewer 2: ```suggestion
You may think we have had to swallow some bitter pills by moving away from an interpreted language, but it has also given us a lot. Most importantly, the rigid structure means that we can execute the whole data pipeline as a single SQL query per layer on the backend. This means that if you want to create a bar plot of 10 billion transactions you only ever fetch the count values for each bar from your data warehouse, not the 10 billion rows of data. The same is true for more complicated layer types such as boxplots and density plots.
```

I don't know that form of the saying. I'd say something like "a bitter pill to swallow".

Reviewer: I think it might be 'bite the bullet', but I think it might be an Americanism due to gun culture yonder

Reviewer: It has a name! TIL: ✨ Malaphor ✨ 
https://malaphors.com/2012/12/31/its-time-to-swallow-the-bullet/
The whole website is like 'A beginner's guide to Euro English'

Reviewer 2: Ah! Either own their own is fine :)

> You may think we have had to swallow some painful bullets by moving away from an interpreted language, but it has also given us a lot. Most importantly, the rigid structure means that we can execute the whole data pipeline as a single SQL query per layer on the backend. This means that if you want to create a bar plot of 10 billion transactions you only ever fetch the count values for each bar from your data warehouse, not the 10 billion rows of data. The same is true for more complicated layer types such as boxplots and density plots.

Reviewer 2: > you only ever fetch the count values for each bar from your data warehouse, not the 10 billion rows of data.

I feel like this is important! This is quite different to how stat calculations normally work in ggplot2 and plotnine, right? If we are to de-emphasise the malicious code angle, we should emphasise this more in its place. It would be nice to see this point expanded on in it's own section, perhaps with a diagram (or humorous image, if we can find one that works) that demonstrates turning huge chunks of data into just a few data points.

### We are now wise beyond our years

![](age.gif)

18 years of ggplot2 development and maintenance also means 18 years of thinking about data visualization syntax, use, and design. I am not a boastful person but I do believe that gives us some expert knowledge on the subject matter. However, not all of this knowledge can be poured back into ggplot2. There are decisions and expectations established many years ago that we have to honor, or at least only challenge very gradually (which we do on occasions).

> ggsql is a blank slate. Not only in the sense that we are building it from the ground up, but also in that it is build for an environment with no established expectations for a visualization tool. I cannot stress how liberating and invigorating this has felt, and I am positive that this shines through in how ggsql feels to the user.

Reviewer 2: ```suggestion
ggsql is a blank slate. Not only in the sense that we are building it from the ground up, but also in that it is built for an environment with no established expectations for a visualization tool. I cannot stress how liberating and invigorating this has felt, and I am positive that this shines through in how ggsql feels to the user.
```

### Declarative wrangling, declarative visualization

> If you are reading this with no prior knowledge of SQL, here's a very brief recap: SQL is a domain specific language for manipulating relational data stored in one or more tables. The syntax is based on the concept of relational algebra which is a structured way to think about data manipulation operations. The semantics defines a set of modular operations that are declarative rather than functional, allowing the user to declare very powerful and custom manipulations using a well-defined set of operations.

Reviewer: It is weird that we've been talking about SQL since the beginning, but only now introduce it. We should maybe give an layman introductory sentence near the start.

Maybe the same applies to the grammar of graphics paragraph hereafter, but I'm not the most impartial judge of that.

Author: I have moved this part up as number 2

> If you are reading this with no prior knowledge of SQL, here's a very brief recap: SQL is a domain specific language for manipulating relational data stored in one or more tables. The syntax is based on the concept of relational algebra which is a structured way to think about data manipulation operations. The semantics defines a set of modular operations that are declarative rather than functional, allowing the user to declare very powerful and custom manipulations using a well-defined set of operations.

Author: ```suggestion
If you are reading this with no prior knowledge of SQL, here's a very brief recap: SQL is a domain specific language for manipulating relational data stored in one or more tables. The syntax is based on the concept of relational algebra which is a structured way to think about data manipulation operations. The semantics defines a set of modular operations that are declarative rather than functional, allowing the user to compose very powerful and custom manipulations using a well-defined set of operations.
```

> If you are reading this with no prior knowledge of the grammar of graphics, here's a very brief recap: The grammar of graphics is a theoretical deconstruction of the concepts of data visualization into its modular parts. While purely theoretical, tools such as ggplot2 have implemented the idea in practice. The semantics defines a set of modular operations that are declarative rather than functional, allowing the user to declare very powerful and custom visualizations using a well-defined set of operations.

Reviewer: ```suggestion
If you are reading this with no prior knowledge of the grammar of graphics, here's a very brief recap: The grammar of graphics is a theoretical deconstruction of the concepts of data visualization into its modular parts. While purely theoretical, tools such as ggplot2 have implemented the idea in practice. The semantics defines a set of modular operations that are declarative rather than functional, allowing the user to compose very powerful and custom visualizations using a well-defined set of operations.
```
I just like this better

Reviewer 2: We should change it in both paragraphs if it makes sense in both, so they continue to match.

That writing device was wonderfully executed, by the way!

![](epic_handshake.png)

> From the above, slightly hyperbolic, overview it is clear that both SQL and the grammar of graphics share a lot of commonality in their approach to their respective domains. Together they can offer a very powerful and natural solution to the full pipeline from raw data to final visualization.

Reviewer: ```suggestion
From the above, slightly hyperbolic, overview it is clear that both SQL and the grammar of graphics have a lot of commonality in their approach to their respective domains. Together they can offer a very powerful and natural solution to the full pipeline from raw data to final visualization.
```
This bordered a tautology. Alternatively: 'share a lot of similarities'

## Meet ggsql
That was a lot of words about the why. Hopefully you come away from it all with a sense of this whole thing being justified, ready to learn more about it. If so, read on!

### A (simple) example
Let's start with some code even though we haven't introduced the syntax yet. I promise we will go through it below. The example is an adaptation of a visualization created by [Jack Davison](https://jack-davison.github.io) for TidyTuesday.

```{ggsql}
WITH astronauts AS (
  SELECT * FROM (
    SELECT *, ROW_NUMBER() OVER (PARTITION BY name ORDER BY mission_number DESC)   AS rn
    FROM 'astronauts.parquet'
  ) t
  WHERE rn = 1

> )

Reviewer 2: You might prefer the slightly shorter,

```
WITH astronauts AS (
  SELECT * FROM 'astronauts.parquet'
  QUALIFY ROW_NUMBER() OVER (PARTITION BY name ORDER BY mission_number DESC) = 1
)
[...]
```

which Claude claims is the same

At the topmost level there are two parts to an SQL query: The SQL query, and the visualization query. The SQL query is anything from the beginning to the `VISUALIZE` clause. It is your standard SQL, and it accepts anything your backend accepts (in this blog post we use a DuckDB backend). The result of the query is funnelled directly into the visualization rather than being returned as a table like you'd normally expect. It should be noted that the SQL query part is optional. If your data is already in the right shape for plotting you can skip it and instead name the source directly in the `VISUALIZE` clause:

```ggsql
VISUALIZE year_of_selection AS x, year_of_mission AS y FROM 'astronauts.parquet'
```

Since the point of this post is not to teach you SQL we won't spend much more time discussing the SQL query part. The main take away is that everything before the `VISUALIZE` clause is pure SQL, any resulting table is automatically used by your visualization, and any table or CTE created there is available for referencing in the visualization query.

> Now, let's look at the new part — everything from `VISUALIZE` and onwards. The visualization query *must* begin with `VISUALIZE` (or `VISUALISE` for those who prefer UK spelling). It can stand on its own or, as we do here, have one or more mappings which will become defaults for every subsequent layer. A mapping is like a `SELECT` where you alias columns to a visual properties (called *aesthetics* in the grammar of graphics). In the visualization above we say that the age column holds the values used for `x` (position along the x axis) and the category column holds the values used for `fill` (the fill color of the entity). We do not say anything about how to draw it yet. Mappings are purely for relating data to abstract visual properties.

Reviewer: ```suggestion
Now, let's look at the new part — everything from `VISUALIZE` and onwards. `VISUALIZE` marks the end of the SQL query and the beginning of the visualization query (or `VISUALISE` for those who prefer UK spelling). It can stand on its own or, as we do here, have one or more mappings which will become defaults for every subsequent layer. Mappings are purely for relating data to abstract visual properties. A mapping is like a `SELECT` where you alias columns to a visual properties (called *aesthetics* in the grammar of graphics). In the visualization above we say that the age column holds the values used for `x` (position along the x axis) and the category column holds the values used for `fill` (the fill color of the entity). We do not say anything about how to draw it yet. 
```

> Following the `VISUALIZE` query we have a `DRAW` query. `DRAW` is how we add layers to our visualization. There are a large selection of different layer types in ggsql. Some are straightforward (e.g. `point` for drawing a scatterplot) and some are more involved like `histogram` (which we use here) requiring calculation of derived statistics like binned count. A visualization can have any number of layers and they will be rendered in sequence as they are defined. `DRAW` has a sibling clause called `PLACE`. It is used for annotation and works like `DRAW` except it doesn't get data from a table but rather as provided literal values. It follows that our visualization above contains three layers: A histogram layer showing data from our table, a rule layer showing precomputed mean values for each category, and a text layer adding context to the visualization. 

Reviewer: Is '`DRAW` query' the right phrasing? Should it be 'statement' or 'clause' instead of 'query'?

Reviewer 2: Query definitely feels wrong to me too, though I'm not sure if we ever settled on statement vs clause. I think either would be fine.

> Following the `VISUALIZE` query we have a `DRAW` query. `DRAW` is how we add layers to our visualization. There are a large selection of different layer types in ggsql. Some are straightforward (e.g. `point` for drawing a scatterplot) and some are more involved like `histogram` (which we use here) requiring calculation of derived statistics like binned count. A visualization can have any number of layers and they will be rendered in sequence as they are defined. `DRAW` has a sibling clause called `PLACE`. It is used for annotation and works like `DRAW` except it doesn't get data from a table but rather as provided literal values. It follows that our visualization above contains three layers: A histogram layer showing data from our table, a rule layer showing precomputed mean values for each category, and a text layer adding context to the visualization. 

Reviewer: ```suggestion
Following the `VISUALIZE` query we have a `DRAW` query. `DRAW` is how we add layers to our visualization. There is a large selection of different layer types in ggsql. Some are straightforward: e.g. `point` for drawing a scatterplot. Some are more involved: `histogram` (which we use here) requires calculation of derived statistics like binned count. A visualization can have any number of layers and layers will be rendered in the sequence they are defined. `DRAW` has a sibling clause called `PLACE`. It is used for annotation and works like `DRAW` except it doesn't get data from a table but rather as provided literal values. It follows that our visualization above contains three layers: A histogram layer showing data from our table, a rule annotation layer showing precomputed mean values for each category, and a text annotation layer adding context to the visualization. 
```

> Following the `VISUALIZE` query we have a `DRAW` query. `DRAW` is how we add layers to our visualization. There are a large selection of different layer types in ggsql. Some are straightforward (e.g. `point` for drawing a scatterplot) and some are more involved like `histogram` (which we use here) requiring calculation of derived statistics like binned count. A visualization can have any number of layers and they will be rendered in sequence as they are defined. `DRAW` has a sibling clause called `PLACE`. It is used for annotation and works like `DRAW` except it doesn't get data from a table but rather as provided literal values. It follows that our visualization above contains three layers: A histogram layer showing data from our table, a rule layer showing precomputed mean values for each category, and a text layer adding context to the visualization. 

Reviewer: Should maybe mention that a layer can display groups simultaneously. This may sound obvious to us but people may have the base R mindset that you have to loop `polygon()` to draw multiple polygons.

> After the `DRAW` and `PLACE` clauses we have a `SCALE` clause. This clause controls how data values are translated into values that are meaningful for the aesthetic (in our case the category column holds the strings "Age at mission" and "Age at selection" which doesn't in itself translate to a color value). The query `SCALE fill TO accent` tells ggsql to use the "accent" color palette when converting the values mapped to fill into actual colors. Scales can be used for much more, like applying transformations to continuous data, defining break points, and setting specific scale types (like e.g. ordinal or binned).

Reviewer: ```suggestion
After the `DRAW` and `PLACE` clauses we have a `SCALE` clause. This clause controls how data values are translated into values that are meaningful for the aesthetic. In our case, the `category` column holds the strings "Age at mission" and "Age at selection" which doesn't in itself translate to a color value. The query `SCALE fill TO accent` tells ggsql to use the "accent" color palette when converting the values mapped to fill to actual colors. Scales can be used for much more, like applying transformations to continuous data, defining break points, and setting specific scale types (like ordinal or binned).
```

The last clause in our visual query is `LABEL` which allows us to add or modify various text labels like title, subtitle, and axis and legend titles. 

### Stepping back
That was a mouthful. But there are two very silvery linings to it all:

1. You now know the most important aspects of the syntax (there are more, of course, but you can grow into that)
2. Many visualization queries will be much simpler

To support the last claim, let's make a boxplot of astronaut birth year split by sex:

```{ggsql}
VISUALIZE sex AS x, year_of_birth AS y FROM 'astronauts.parquet'
DRAW boxplot
```

> That's much shorter but still, if you are coming from a different plotting system you may even think this is overly verbose (e.g. compared to something like `boxplot(astronauts.sex, astronauts.year_of_birth)`). Yes, it is longer, but it is also more structured, composable, and self-descriptive. These features (which are a direct result of its grammar of graphics lineage) means that both you and your future llm coding buddy will have an easier time internalizing the workings of *all* types of plots that can be made. The 18 years of dominance of ggplot2 (which shares these feature) in the R ecosystem is a testament to this.

Reviewer 2: ```suggestion
That's much shorter but still, if you are coming from a different plotting system you may even think this is overly verbose (e.g. compared to something like `boxplot(astronauts.sex, astronauts.year_of_birth)`). Yes, it is longer, but it is also more structured, composable, and self-descriptive. These features (which are a direct result of its grammar of graphics lineage) means that both you and your future LLM coding buddy will have an easier time internalizing the workings of *all* types of plots that can be made. The 18 years of dominance of ggplot2 (which shares these features) in the R ecosystem is a testament to this.
```

As an example, let's change the above plot to instead show the same relationship as a jittered scatterplot.

```{ggsql}
VISUALIZE sex AS x, year_of_birth AS y FROM 'astronauts.parquet'
DRAW point
  SETTING position => 'jitter'
```

Or perhaps the jitter follows the distribution of the data so it doubles as a violin plot:

```{ggsql}
VISUALIZE sex AS x, year_of_birth AS y FROM 'astronauts.parquet'
DRAW point
  SETTING position => 'jitter', distribution => 'density'
```

As you can see the syntax and composable nature makes visualization iteration very ergonomic, something that is extremely valuable in both explorative analyses and visualization design.

## Behind the curtains

> With a passing understanding of how to use ggsql, let's discuss what goes on behind the scene. At its core ggsql is a modular rust library that is split into a Reader module, a Plot module, and a Writer module. Both the Reader and Writer can be swapped out, leaving the Plot module as the only constant.

Reviewer 2: ```suggestion
With a passing understanding of how to use ggsql, let's discuss what goes on behind the scenes. At its core ggsql is a modular rust library that is split into a Reader module, a Plot module, and a Writer module. Both the Reader and Writer can be swapped out, leaving the Plot module as the only constant.
```

#### The Reader

> This module is responsible for communicating with your data backend, i.e. your SQL database. While this make it sound like you need to have a complicated database setup in order to use ggsql, this is not the case. While ggsql ships with an [ODBC](https://wikipedia.org/wiki/Open_Database_Connectivity) and [Snowflake](https://www.snowflake.com/) reader, the default reader is [DuckDB](https://duckdb.org) which allows you to ingest e.g. parquet and csv files directly (no DuckDB installation necessary) as we have done above, as well as talk to any DuckDB database you may have set up.

Reviewer: ```suggestion
This module is responsible for communicating with your data backend, i.e. your SQL database. While this makes it sound like you need to have a complicated database setup in order to use ggsql, this is not the case. While ggsql ships with an [ODBC](https://wikipedia.org/wiki/Open_Database_Connectivity) and [Snowflake](https://www.snowflake.com/) reader, the default reader is [DuckDB](https://duckdb.org). DuckDB allows you to ingest e.g. parquet and csv files directly (no installation necessary) as we have done above, as well as talk to any DuckDB database you may have set up.
```

Reviewer 2: > as well as talk to any DuckDB database you may have set up.

I think it is quite strange to have a remote DuckDB database, it is really designed to be in-process.

Instead we should take the opportunity to mention in passing that you can connect to different remote SQL servers via DuckDB extensions, e.g. https://duckdb.org/docs/current/core_extensions/postgres

#### The Plot

> The Reader module communicates with the Plot module in two ways. It tells the Plot module about it's particular flavor of SQL and allows it to execute queries on it, returning the resulting data. The Plot module on the other hand is responsible for reading the visualize query and parse it into a plot representation that can be handed off to the Writer. Part of this is to construct a single query for each layer in the visualization (each `DRAW` query) that returns the data needed for that layer. All computations necessary for the layer is part of this query and are thus handled by the backend. For instance, the query for the boxplot layer includes logic for calculating quantiles etc, all happening in the database without the need for materializing the raw data.

Reviewer: ```suggestion
The Reader module communicates with the Plot module in two ways. It tells the Plot module about it's particular flavor of SQL and allows it to execute queries on it, returning the resulting data. On the other hand, the Plot module is responsible for reading the visualize query and parse it into a plot representation that can be handed off to the Writer. Part of this is to construct a single query for each layer in the visualization (each `DRAW` query) that returns the data needed for that layer. All computations necessary for the layer is part of this query and are thus handled by the backend. For instance, the query for the boxplot layer includes logic for calculating quantiles etc, all happening in the database without the need for materializing the raw data.
```

#### The Writer

> This module is responsible for the output. It gets the structured visualization representation from the Plot module and converts it into the actual visual representation of it. Currently we only have a [Vegalite](https://vega.github.io/vega-lite/) writer but we are planning to expand this very soon to shed our reliance on Vegalite idioms.

Reviewer: ```suggestion
This module is responsible for the output. It gets the structured representation of the visualization from the Plot module and converts it into the actual visual representation of it. Currently we only have a [Vegalite](https://vega.github.io/vega-lite/) writer but we are planning to expand this very soon to shed our reliance on Vegalite idioms.
```

### Tooling
More important than how it works underneath is perhaps how you use it. Recognizing that people may wish to use it in different ways we provide ggsql in a number of different distributions:

#### Jupyter kernel
As part of the standard installation we provide a ggsql [Jupyter](https://jupyter.org) kernel. This allows you to use ggsql directly within Jupyter notebooks and [Quarto](https://quarto.org) documents, both for executing SQL and ggsql queries. It maintains a persistent connection throughout the document so that visual queries can reference tables and views created in earlier blocks.

#### CLI
The second part of the standard installation is a command line tool for executing queries from the command line, either by providing the query in a file or directly in the call. It also provides parsing and validation tools.

#### VS Code + Positron extension

> The [ggsql VS Code extension](https://open-vsx.org/extension/ggsql/ggsql) provides a tailored ggsql experience for [Visual Studio Code](https://code.visualstudio.com) and in particular for [Positron](https://positron.posit.co). In Positron you will get a ggsql console where you can write and execute queries, with the result appear in the plot pane along with your R and Python plots. You also get cmd+enter support for sending commands from a ggsql script to the console for execution. Lastly, in Positron we interface with the Connection pane allowing you to hook up ggsql to any database you have there and immediately begin visualizing the data in your data warehouse. 

Reviewer: ```suggestion
The [ggsql VS Code extension](https://open-vsx.org/extension/ggsql/ggsql) provides a tailored ggsql experience for [Visual Studio Code](https://code.visualstudio.com) and in particular for [Positron](https://positron.posit.co). In Positron you will get a ggsql console where you can write and execute queries, with the result appearing in the plot pane along with your R and Python plots. You also get cmd+enter support for sending commands from a ggsql script to the console for execution. Lastly, in Positron we interface with the Connection pane allowing you to hook up ggsql to any database you have there and immediately begin visualizing the data in your data warehouse. 
```
cmd + enter is a mac-ism and users might re-bind that

Reviewer 2: > cmd + enter is a mac-ism and users might re-bind that

I think "Send to Console support" is a clear alternative.

> The [ggsql VS Code extension](https://open-vsx.org/extension/ggsql/ggsql) provides a tailored ggsql experience for [Visual Studio Code](https://code.visualstudio.com) and in particular for [Positron](https://positron.posit.co). In Positron you will get a ggsql console where you can write and execute queries, with the result appear in the plot pane along with your R and Python plots. You also get cmd+enter support for sending commands from a ggsql script to the console for execution. Lastly, in Positron we interface with the Connection pane allowing you to hook up ggsql to any database you have there and immediately begin visualizing the data in your data warehouse. 

Reviewer 2: A Positron screenshot might be nice here, we could show a `.gsql` file, a Quarto document, some code in the console pane, and the rendered Quarto output on the right. Something like this:

<img width="2129" height="1290" alt="Screenshot 2026-04-15 at 15 32 45" src="https://github.com/user-attachments/assets/7a057fad-0da0-4317-8a8d-3ecd572874b5" />

> The [ggsql VS Code extension](https://open-vsx.org/extension/ggsql/ggsql) provides a tailored ggsql experience for [Visual Studio Code](https://code.visualstudio.com) and in particular for [Positron](https://positron.posit.co). In Positron you will get a ggsql console where you can write and execute queries, with the result appear in the plot pane along with your R and Python plots. You also get cmd+enter support for sending commands from a ggsql script to the console for execution. Lastly, in Positron we interface with the Connection pane allowing you to hook up ggsql to any database you have there and immediately begin visualizing the data in your data warehouse. 

Author: ```suggestion
The [ggsql VS Code extension](https://open-vsx.org/extension/ggsql/ggsql) provides a tailored ggsql experience for [Visual Studio Code](https://code.visualstudio.com) and in particular for [Positron](https://positron.posit.co). In Positron you will get a ggsql console where you can write and execute queries, with the result appear in the plot pane along with your R and Python plots. You also get support for sending commands from a ggsql script to the console for execution. Lastly, in Positron we interface with the Connection pane allowing you to hook up ggsql to any database you have there and immediately begin visualizing the data in your data warehouse. 
```

<img src="epic_handshake.png" alt="Epic handshake meme showing SQL and the grammar of graphics agreeing on declarative, composable operations" />

From the above, slightly hyperbolic, overview it is clear that both SQL and the grammar of graphics have a lot of commonality in their approach to their respective domains. Together they can offer a very powerful and natural solution to the full pipeline from raw data to final visualization.

### No runtime, no problem
Why does it matter that [ggplot2](https://ggplot2.tidyverse.org/) and [plotnine](https://plotnine.org) requires R and Python installed respectively? There are clear benefits to a single, focused executable to handle data visualization:

> * Embedding a small executable in other tools are much easier than bundling R/Python (or requiring them to be installed)

Reviewer 2: ```suggestion
* Embedding a small executable in other tools is much easier than bundling R/Python (or requiring them to be installed)
```

* Clear scope and need makes it easier to sandbox and prevent malicious code execution (either deliberately or in error)

Both of the above points make ggsql a much more compelling option for integrating into tools such as AI agents assisting you in data analysis, or code based reporting tools that may execute code in different environments.

> You may think we have had to swallow some bitter pills by moving away from an interpreted language, but it has also given us a lot. Most importantly, the rigid structure means that we can execute the whole data pipeline as a single SQL query per layer on the backend. This means that if you want to create a bar plot of 10 billion transactions you only ever fetch the count values for each bar from your data warehouse, not the 10 billion rows of data. The same is true for more complicated layer types such as boxplots and density plots. This is in stark contrast to the situation in most visualization tools that first materializes the data (potentially chocking on the size), then perform the necessary computations on it, before proceeding to plot it.

Reviewer 2: ```suggestion
You may think we have had to swallow some bitter pills by moving away from an interpreted language, but it has also given us a lot. Most importantly, the rigid structure means that we can execute the whole data pipeline as a single SQL query per layer on the backend. This means that if you want to create a bar plot of 10 billion transactions you only ever fetch the count values for each bar from your data warehouse, not the 10 billion rows of data. The same is true for more complicated layer types such as boxplots and density plots. This is in stark contrast to the situation in most visualization tools that first materializes the data (potentially choking on the size), then perform the necessary computations on it, before proceeding to plot it.
```

```

Since the point of this post is not to teach you SQL we won't spend much more time discussing the SQL query part. The main take away is that everything before the `VISUALIZE` clause is pure SQL, any resulting table is automatically used by your visualization, and any table or CTE created there is available for referencing in the visualization query.

Now, let's look at the new part — everything from `VISUALIZE` and onwards. `VISUALIZE` marks the end of the SQL query and the beginning of the visualization query (or `VISUALISE` for those who prefer UK spelling). It can stand on its own or, as we do here, have one or more mappings which will become defaults for every subsequent layer. Mappings are purely for relating data to abstract visual properties. A mapping is like a `SELECT` where you alias columns to a visual properties (called *aesthetics* in the grammar of graphics). In the visualization above we say that the age column holds the values used for `x` (position along the x axis) and the category column holds the values used for `fill` (the fill color of the entity). We do not say anything about how to draw it yet. 

> Following the `VISUALIZE` query we have a `DRAW` clause. `DRAW` is how we add layers to our visualization. There is a large selection of different layer types in ggsql. Some are straightforward: e.g. `point` for drawing a scatterplot. Some are more involved: `histogram` (which we use here) requires calculation of derived statistics like binned count. A visualization can have any number of layers and layers will be rendered in the sequence they are defined. `DRAW` has a sibling clause called `PLACE`. It is used for annotation and works like `DRAW` except it doesn't get data from a table but rather as provided literal values. It follows that our visualization above contains three layers: A histogram layer showing data from our table, a rule annotation layer showing precomputed mean values for each category, and a text annotation layer adding context to the visualization. It is worth mentioning that a layer does not correspond to a single graphical entity. Like with the text layer above, every layer can render multiple separate entities of its type so there is no need have e.g. 3 line layers to render line plots for 3 different categories.

Reviewer 2: ```suggestion
Following the `VISUALIZE` query we have a `DRAW` clause. `DRAW` is how we add layers to our visualization. There is a large selection of different layer types in ggsql. Some are straightforward: e.g. `point` for drawing a scatterplot. Some are more involved: `histogram` (which we use here) requires calculation of derived statistics like binned count. A visualization can have any number of layers and layers will be rendered in the sequence they are defined. `DRAW` has a sibling clause called `PLACE`. It is used for annotation and works like `DRAW` except it doesn't get data from a table but rather as provided literal values. It follows that our visualization above contains three layers: A histogram layer showing data from our table, a rule annotation layer showing precomputed mean values for each category, and a text annotation layer adding context to the visualization. It is worth mentioning that a layer does not correspond to a single graphical entity. Like with the text layer above, each layer can render multiple separate entities of its type so there is no need to have e.g. 3 line layers to render line plots for 3 different categories.
```

After the `DRAW` and `PLACE` clauses we have a `SCALE` clause. This clause controls how data values are translated into values that are meaningful for the aesthetic. In our case, the `category` column holds the strings "Age at mission" and "Age at selection" which doesn't in itself translate to a color value. The clause `SCALE fill TO accent` tells ggsql to use the "accent" color palette when converting the values mapped to fill to actual colors. Scales can be used for much more, like applying transformations to continuous data, defining break points, and setting specific scale types (like ordinal or binned).

The last clause in our visual query is `LABEL` which allows us to add or modify various text labels like title, subtitle, and axis and legend titles. 

### Stepping back
That was a mouthful. But there are two very silvery linings to it all:

1. You now know the most important aspects of the syntax (there are more, of course, but you can grow into that)

> 2. Many visualization queries will be much simpler

Reviewer 2: ```suggestion
2. Many visualization queries will be much simpler than the one above
```

As an example, let's change the above plot to instead show the same relationship as a jittered scatterplot.

```{ggsql}
VISUALIZE sex AS x, year_of_birth AS y FROM 'astronauts.parquet'
DRAW point
  SETTING position => 'jitter'
```

Or perhaps the jitter follows the distribution of the data so it doubles as a violin plot:

```{ggsql}
VISUALIZE sex AS x, year_of_birth AS y FROM 'astronauts.parquet'
DRAW point
  SETTING position => 'jitter', distribution => 'density'
```

As you can see the syntax and composable nature makes visualization iteration very ergonomic, something that is extremely valuable in both explorative analyses and visualization design.

## Behind the curtains

> With a passing understanding of how to use ggsql, let's discuss what goes on behind the scenes. At its core ggsql is a modular rust library that is split into a Reader module, a Plot module, and a Writer module. Both the Reader and Writer can be swapped out, leaving the Plot module as the only constant.

Reviewer 2: ```suggestion
With a passing understanding of how to use ggsql, let's discuss what goes on behind the scenes. At its core ggsql is a modular Rust library that is split into a Reader module, a Plot module, and a Writer module. Both the Reader and Writer can be swapped out, leaving the Plot module as the only constant.
```

Rust should be written as a proper noun, possibly there are other instances I've missed.

#### The Reader

> This module is responsible for communicating with your data backend, i.e. your SQL database. While this makes it sound like you need to have a complicated database setup in order to use ggsql, this is not the case. While ggsql ships with an [ODBC](https://wikipedia.org/wiki/Open_Database_Connectivity) and [Snowflake](https://www.snowflake.com/) reader, the default reader is [DuckDB](https://duckdb.org). DuckDB allows you to ingest e.g. parquet and csv files directly (no installation necessary) as we have done above, as well as talk to remote SQL databases through its rich extension ecosystem.

Reviewer 2: ```suggestion
This module is responsible for communicating with your data backend, i.e. your SQL database. While this makes it sound like you need to have a complicated database setup in order to use ggsql, this is not the case. While ggsql ships with an [ODBC](https://wikipedia.org/wiki/Open_Database_Connectivity) and [Snowflake](https://www.snowflake.com/) reader, the default reader is [DuckDB](https://duckdb.org). DuckDB allows you to ingest e.g. parquet and csv files directly (no installation necessary) as we have done above, as well as talk to remote SQL databases through its rich (extension ecosystem)[https://duckdb.org/docs/current/core_extensions/overview].
```

Might be worth linking?
